src/core/server/services/users/index.ts

```diff
@@ -11,14 +12,105 @@ import {
   User,
 } from "talk-server/models/user";
 
+const USERNAME_REGEX = new RegExp(/^[a-zA-Z0-9_]+$/);
+const USERNAME_MAX_LENGTH = 30;
+
+/**
+ * validateUsername will validate that the username is valid. Current
+ * implementation uses a RegExp statically, future versions will expose this as
+ * configuration.
+ *
+ * @param tenant tenant where the User is associated with
+ * @param username the username to be tested
+ */
+function validateUsername(tenant: Tenant, username: string) {
+  // TODO: replace these static regex/length with database options in the Tenant eventually
+
+  if (!USERNAME_REGEX.test(username)) {
+    throw new Error("username contained illegal characters");
+  }
+
+  if (username.length > USERNAME_MAX_LENGTH) {
+    throw new Error("username exceeded maximum length");
+  }
+}
+
+/**
+ * validatePassword will validate that the password is valid. Current
+ * implementation uses a length statically, future versions will expose this as
+ * configuration.
+ *
+ * @param tenant tenant where the User is associated with
+ * @param password the password to be tested
+ */
+function validatePassword(tenant: Tenant, password: string) {
+  // TODO: replace these static length with database options in the Tenant eventually
+  if (password.length < 8) {
+    throw new Error("password is too short");
+  }
+}
+
+const EMAIL_REGEX = new RegExp(/^\S+@\S+.\S+$/);
```

Comment:
The client currently uses a different REGEX: see `src/core/client/framework/lib/validation.tsx`. Should we maybe expose the MIN + MAX_LENGTH constants and the REGEXs in `talk-common` and reuse them in both? At least until we get to better error handling.